include/exadg/fluid_structure_interaction/precice/precice_adapter.h

```diff
@@ -0,0 +1,380 @@
+  std::map<std::string, std::shared_ptr<CouplingSurface<dim, data_dim, VectorizedArrayType>>>
+    writer;
+  // We restrict the reader to be of type ExaDGSurface for the moment, as all other choices don't
+  // make sense
+  std::map<std::string, std::shared_ptr<ExaDGSurface<dim, data_dim, VectorizedArrayType>>> reader;
+
+  // Container to store time dependent data in case of an implicit coupling
+  std::vector<VectorType> old_state_data;
+};
+
+
+
+template<int dim, int data_dim, typename VectorType, typename VectorizedArrayType>
+template<typename ParameterClass>
+Adapter<dim, data_dim, VectorType, VectorizedArrayType>::Adapter(const ParameterClass & parameters,
+                                                                 MPI_Comm               mpi_comm)
+
+{
+  precice = std::make_shared<precice::SolverInterface>(parameters.participant_name,
+                                                       parameters.config_file,
+                                                       Utilities::MPI::this_mpi_process(mpi_comm),
+                                                       Utilities::MPI::n_mpi_processes(mpi_comm));
+
+  AssertThrow(dim == precice->getDimensions(), ExcInternalError());
+  AssertThrow(dim > 1, ExcNotImplemented());
+}
+
+
+
+template<int dim, int data_dim, typename VectorType, typename VectorizedArrayType>
+void
+Adapter<dim, data_dim, VectorType, VectorizedArrayType>::add_write_surface(
+  const types::boundary_id                                            dealii_boundary_surface_id,
+  const std::string &                                                 mesh_name,
+  const std::vector<std::string> &                                    write_data_names,
+  const WriteDataType                                                 write_data_type,
+  std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>> data,
+  const unsigned int                                                  dof_index,
+  const unsigned int                                                  quad_index)
+{
+  // Check, if we already have such an interface
+  const auto found_reader = reader.find(mesh_name);
+
+  if(found_reader != reader.end())
+  {
+    // Duplicate the shared pointer
+    writer.insert({mesh_name, found_reader->second});
+  }
+  else if(write_data_type == WriteDataType::values_on_dofs)
+  {
+    writer.insert({mesh_name,
+                   std::make_shared<DoFSurface<dim, data_dim, VectorizedArrayType>>(
+                     data, precice, mesh_name, dealii_boundary_surface_id, dof_index)});
+  }
+  else
+  {
+    Assert(write_data_type == WriteDataType::values_on_q_points ||
+             write_data_type == WriteDataType::normal_gradients_on_q_points,
+           ExcNotImplemented());
+    writer.insert({mesh_name,
+                   std::make_shared<QuadSurface<dim, data_dim, VectorizedArrayType>>(
+                     data, precice, mesh_name, dealii_boundary_surface_id, dof_index, quad_index)});
+  }
```

Comment:
I would move this logic to the calling function. As far as I see it, the adapter class is not limited to surfaces but this introduces such a dependency.

Reply:
I think this might be a good idea, but it requires certainly a broader rethinking of the class and interface design. I don't have a silver bullet solution right now. Are you simply thinking of write access to the `writer` and `reader` in order to fill them from the outside of this class?

Reply:
Let's postpone this!